features/storage/kvstore/conf/kv_config.cpp

```diff
@@ -689,57 +548,71 @@ MBED_WEAK BlockDevice *get_other_blockdevice()
     return NULL;
 }
 
-int _storage_config_TDB_INTERNAL()
+int _create_internal_tdb(BlockDevice **internal_bd, KVStore **internal_tdb, bd_size_t *size, bd_addr_t *start_address)
 {
-#if COMPONENT_FLASHIAP
-    bd_size_t internal_size = MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_SIZE;
-    bd_addr_t internal_start_address = MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_BASE_ADDRESS;
     int ret;
 
-    if (internal_size == 0 && internal_start_address == 0) {
+    //Get the default address and size for the TDBStore
+    if (*size == 0 && *start_address == 0) {
         //Calculate the block device size and start address in case default values are used.
-        ret = _get_flashiap_bd_default_addresses_tdb_internal(&internal_start_address, &internal_size);
+        ret = TDBStore::get_default_flash_addresses(start_address, size);
         if (ret != MBED_SUCCESS) {
-            return ret;
+            return MBED_ERROR_FAILED_OPERATION;
         }
     }
 
-    //Get internal memory FLASHIAP block device.
-    kvstore_config.internal_bd = GET_BLOCKDEVICE(INTERNAL_BLOCKDEVICE_NAME, internal_start_address, internal_size);
-    if (kvstore_config.internal_bd == NULL) {
+    //Create internal FLASHIAP block device
+    *internal_bd = GET_BLOCKDEVICE(INTERNAL_BLOCKDEVICE_NAME, *start_address, *size);
+    if (*internal_bd == NULL) {
         tr_error("KV Config: Fail to get internal BlockDevice.");
-        return MBED_ERROR_FAILED_OPERATION;
+        return MBED_ERROR_FAILED_OPERATION ;
     }
 
-
-    ret = kvstore_config.internal_bd->init();
+    // Initialize internal block device
+    ret = (*internal_bd)->init();
     if (ret != MBED_SUCCESS) {
         tr_error("KV Config: Fail to init internal BlockDevice.");
-        return MBED_ERROR_FAILED_OPERATION;
+        return MBED_ERROR_FAILED_OPERATION ;
     }
 
-    //Check that internal flash has 2 or more sectors
-    if (_calculate_blocksize_match_tdbstore(kvstore_config.internal_bd) != MBED_SUCCESS) {
+    //Check if TDBStore has at least 2 sector.
```

Comment:
Why would 10 pages work instead of a full sector? I think TDBSTore uses the flash sectors to do the garbage collection, in sort of A vs. B, one of them being active and the other one as the spare?

Reply:
Yes, that matches my understanding of TDBStore's garbage collection - it divides the available flash area in half and alternates between the two to do copying GC.
The problem is that the default approach breaks down when the sector size and page sizes are the same (or close), because many of the KVStore operations consume an entire page. Based on our analysis of the KVStore greentea tests as a proxy for typical usage, the minimum page count appeared to be around 10 i.e. 5 pages in active use, 5 in reserve for GC (or 2 sectors if that is larger, so that GC continues to work). If there's a different number of pages that more directly corresponds to something intrinsic in the driver I'd be happy to go with that instead.

Reply:
@VeijoPesonen, wouldn't expanding this check produce a backwards compatibility concern? Or does this qualify as a situation where it would probably have always failed, just in a less explicit manner, and therefore it is okay to add a new error condition?

Reply:
@VeijoPesonen, can you please confirm that the expanded check is not a backwards compatibility concern?

Reply:
Hi, was browsing some PRs randomly, sorry for jumping in.

The number of pages TDBStore needs depends on the number of KV pairs+1. So for 4 KV pairs it needs 5 pages. (Actually it may need 2 pages per KV pair, not sure).

Unfortunately I think this means you can't make this check unless you know the number of KV pairs beforehand.

Reply:
I agree that you can't compute a minimum that is guaranteed to be enough unless you know how much is intended to be stored. However, there is a bare minimum size (e.g. space for internal bookkeeping, a second sector for garbage collection) below which TDBStore cannot function. That is what this check is intended to achieve. Are you saying that the bare minimum page count calculation should be different?

Reply:
Ah, my understanding was that the theoretical minimum would be 4 pages total (2 for the active master record, 2 for the old master record, where pages = program size'd units). But my understand could be out of date.

Out of curiousity do you know where the other 2x3 pages come from?

Reply:
I not able to determine precisely where that came from.
However, after further consideration I've removed the hard-fail on too few pages given that the exact minimum is application specific. Full reasoning is in the commit message for a3a4eef59939741999101f1eb5b4367f422f9e6e